primitives/trie/src/storage_proof.rs

```diff
@@ -14,65 +14,990 @@
+		}
+	}
+}
+
+#[cfg(feature = "std")]
+impl<E: sp_std::fmt::Display> sp_std::convert::From<sp_std::boxed::Box<E>> for Error {
+	fn from(e: sp_std::boxed::Box<E>) -> Self {
+		// Only trie error is build from box so we use a tiny shortcut here.
+		Error::Trie(format!("{}", e))
+	}
+}
+
+#[cfg(not(feature = "std"))]
+impl<E> sp_std::convert::From<sp_std::boxed::Box<E>> for Error {
+	fn from(_e: sp_std::boxed::Box<E>) -> Self {
+		Error::Trie
+	}
+}
+
+impl sp_std::convert::From<CodecError> for Error {
+	fn from(e: CodecError) -> Self {
+		error(e.what())
+	}
+}
+
+#[cfg(feature = "std")]
+const fn error(message: &'static str) -> Error {
+	Error::Proof(message)
+}
+
+#[cfg(not(feature = "std"))]
+const fn error(_message: &'static str) -> Error {
+	Error::Proof
+}
+
+const fn missing_pack_input() -> Error {
+	error("Packing input missing for proof")
+}
+
+const fn missing_verify_input() -> Error {
+	error("Input missing for proof verification")
+}
+
+const fn no_partial_db_support() -> Error {
+	error("Partial db not supported for this proof")
+}
+
+/// Different kind of proof representation are allowed.
+/// This definition is used as input parameter when producing
+/// a storage proof.
+/// Some kind are reserved for test or internal use and will
+/// not be usable when decoding proof.
+#[repr(u8)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum StorageProofKind {
+	/// Kind for `StorageProof::Flatten`.
+	Flatten = 1,
+
+	/// Kind for `StorageProof::TrieSkipHashes`.
+	TrieSkipHashes = 2,
+
+	/// Kind for `StorageProof::KnownQueryPlanAndValues`.
+	KnownQueryPlanAndValues = 3,
+
+	/// Technical only
+
+	/// Kind for `StorageProof::TrieSkipHashesForMerge`.
+	TrieSkipHashesForMerge = 125,
+
+	/// Testing only indices
+
+	/// Kind for `StorageProof::Full`.
+	Full = 126,
+
+	/// Kind for `StorageProof::TrieSkipHashesFull`.
+	TrieSkipHashesFull = 127,
+}
+
+impl StorageProofKind {
+	/// Decode a byte value representing the storage kind.
+	/// Return `None` if the kind does not exists or is not allowed.
+	#[cfg(test)]
+	pub fn read_from_byte(encoded: u8) -> Option<Self> {
+		Some(match encoded {
+			x if x == StorageProofKind::Flatten as u8 => StorageProofKind::Flatten,
+			x if x == StorageProofKind::TrieSkipHashes as u8 => StorageProofKind::TrieSkipHashes,
+			x if x == StorageProofKind::KnownQueryPlanAndValues as u8
+				=> StorageProofKind::KnownQueryPlanAndValues,
+			x if x == StorageProofKind::Full as u8 => StorageProofKind::Full,
+			x if x == StorageProofKind::TrieSkipHashesFull as u8 => StorageProofKind::TrieSkipHashesFull,
+			x if x == StorageProofKind::TrieSkipHashesForMerge as u8
+				=> StorageProofKind::TrieSkipHashesForMerge,
+			x if x == StorageProofKind::TrieSkipHashesFull as u8
+				=> StorageProofKind::TrieSkipHashesFull,
+			_ => return None,
+		})
+	}
+
+	/// Decode a byte value representing the storage kind.
+	/// Return `None` if the kind does not exists or is not allowed.
+	#[cfg(not(test))]
```

Comment:
That's really confusing, why do the variatns exist in the first place if they can't be decoded?